### DISCUSSION TITLE
insert_textbox vs insert_htmlbox

### DISCUSSION THREAD

**Discussion by tenghian** [pymupdf/PyMuPDF#3932] [Looking for help]
Hi, when i use insert_textbox get correct font, but insert_htmlbox get different result, may i know how to change args in insert_htmlbox? thanks
[jp_1.pdf](https://github.com/user-attachments/files/17317074/jp_1.pdf)
[jp_2.pdf](https://github.com/user-attachments/files/17317075/jp_2.pdf)

#pymupdf 1.24.8

`import pymupdf

fname = "r:/jp_1.pdf"
new_file = "r:/jp_2.pdf"

doc = pymupdf.open(fname) # open a document
page = doc[0]

pfont = page.get_fonts(full=True)
print(pfont)
#[(11, 'ttf', 'TrueType', 'ABKYFM+HelveticaNeue-Bold', 'TT0', 'WinAnsiEncoding', 0),
# (14, 'ttf', 'TrueType', 'ABFMUM+HelveticaNeue', 'TT1', 'WinAnsiEncoding', 0)]

fontbuffer = doc.extract_font(11)[-1]
page.insert_font(fontname="TT0", fontbuffer=fontbuffer)\

#-> font:HelveticaNeue-Bold (display correct font)
rect1 = pymupdf.Rect((100.0, 550.0, 300.0, 600.0))
page.insert_textbox(rect1, buffer="Embracing", fontsize=24, fontname="TT0")

#-> font:CharisSIL (insert_htmlbox cannot detect HelveticaNeue-Bold)
rect2 = pymupdf.Rect((100.0, 600.0, 300.0, 650.0))
page.insert_htmlbox(rect2, 'Embracing', css="* {font-family: HelveticaNeue-Bold;font-size:24px;}")

doc.save(new_file)
doc.close()

`

**Comment by JorjMcKie** [comment]
All you have to do is using CSS source (= styling information).
If you let MuPDF (the ultimate font selection logic behind everything `Story`-based like htmlbox), it will decide character-by-character, and the default is the "serif" font Charis.
If the character set of that is exhausted, Noto Serif is used - but previous decisions for Charis are of course not reverted in a retrograde manner.
So you can use `css="* [font-family: sans-serif;}", which will lead to using the Noto Sans fonts (regular, bold, italic, ...) and no Charis, Noto Serif, etc.
Or you specify one or more fonts from package pymupdf-fonts or some arbitrary choice of your own favorite fonts.
Look at the recipes for `insert_htmlbox` to see some more examples, among which are also file-based ones.

In any Case, MuPDF **_never_** looks at the fonts **_intalled_** on your machine. Either you accept the default behavior or you must actively provide fonts via file spec or memory equivalents.

**Comment by tenghian** [comment]
insert_textbox can read fonts within the page, how can I make insert_htmlbox do the same?

Is there a sample code to read fonts from memory? thanks

**Reply by JorjMcKie** [reply]
Method `page.insert_textbox()` does **_not_** "read" fonts from the PDF!
It is you who must first extract a font from the PDF and then make it available again via `page.insert_font()` giving it a new `fontname` to make it referenceable for insert_textbox.

A font existing in a PDF can also be made available to `page.insert_htmlbox()`.
Because this method depends on a `Story` object and text styling is supported via HTML and CSS syntax, you must provide font support in that syntax too.
1. Extract existing font as before and create a `Font` object.
2. Create a PyMuPDF `Archive` object that references this `Font` via `font.buffer`.
3. Define CSS source that references the extracted font.
4. Do `page.insert_htmlbox(rect, text, archive=arch, css=css)`

Define Archive and CSS:
```python
arch = pymupdf.Archive(font.buffer, "myfont")  # make font accessible via a name
css = """@font-face {font-family: myfont; src: url(myfont);}
      * {font-family: myfont;}"""
```

As always: a font in a PDF normally is not complete, but a subset of some full font: several / many characters are missing.
If characters are missing in supplied fonts, Story will behave differently than `insert_textbox` ... so you need to experiment a little here.

**Comment by tenghian** [comment]
fname = "r:/jp_1.pdf"
new_file = "r:/jp_2.pdf"

doc = pymupdf.open(fname) # open a document
page = doc[0]

pfont = page.get_fonts(full=True)
print(pfont)

fontbuffer = doc.extract_font(11)[-1]
font = pymupdf.Font(fontname="TT0", fontbuffer=fontbuffer)
page.insert_font(fontname="TT0", fontbuffer=fontbuffer)

arch = pymupdf.Archive(font.buffer, "myfont")  # make font accessible via a name
css = """@font-face {font-family: myfont; src: url(myfont);}
      * {font-family: myfont;}"""

rect2 = pymupdf.Rect((100.0, 600.0, 300.0, 650.0))
page.insert_htmlbox(rect2, 'Embracing', css=css)

rect1 = pymupdf.Rect((100.0, 550.0, 300.0, 600.0))
page.insert_textbox(rect1, buffer="Embracing", fontsize=24, fontname="TT0")

-> MuPDF error: format error: cannot locate font 'myfont' specified by css
I am using the text and font that already exist on the page, and insert_textbox can read it normally.

**Reply by JorjMcKie** [reply]
You did not specify the `archive` parameter in `insert_htmlblox`!

**Comment by JorjMcKie** [comment]
This demo script works as desired:
```python
import pymupdf
from pprint import pprint

doc = pymupdf.open("test-doc-2.pdf")
buff = doc.extract_font(124)[-1]
doc.close()  # original PDF no longer needed

font = pymupdf.Font(fontbuffer=buff)
arch = pymupdf.Archive(font.buffer, "myfont")
css = """ @font-face {font-family: myfont; src: url(myfont);}
* {font-family: myfont;}"""
doc = pymupdf.open()
page = doc.new_page()
page.insert_htmlbox((100, 100, 200, 200), "Text in custom font", css=css, archive=arch)
pprint(page.get_fonts())
```

**Reply by tenghian** [reply]
Thank you. ^_^
